File: python/octoarms_sdk/migration_test.py
from __future__ import annotations

import json
import shutil
import tempfile
import unittest
from pathlib import Path

from octoarms_sdk.migration import migrate


class MigrateTests(unittest.TestCase):
    def test_posts_declared_migration_sql_to_capability_api(self) -> None:
        root = Path(tempfile.mkdtemp(prefix="octoarms-sdk-migrate-"))
        calls: list[dict[str, object]] = []

        try:
            migrations_dir = root / "migrations"
            migrations_dir.mkdir(parents=True, exist_ok=True)
            (migrations_dir / "20260329_01_init.sql").write_text("SELECT 1;\n", encoding="utf-8")

            def fake_request(url: str, method: str, headers: dict[str, str], body: bytes) -> tuple[int, str]:
                calls.append(
                    {
                        "url": url,
                        "method": method,
                        "headers": headers,
                        "body": body.decode("utf-8"),
                    }
                )
                return 200, '{"code":0,"msg":"success"}'

            result = migrate(
                {
                    "task_name": "defillama_tvl_top10",
                    "task_version": "v0.1.0",
                    "migration_version": "20260329_01_init",
                    "capability_endpoint": "http://capability.local",
                    "ephemeral_token": "ephemeral-token",
                    "migration_index": {"20260329_01_init": "migrations/20260329_01_init.sql"},
                    "task_root_dir": str(root),
                },
                request_fn=fake_request,
            )

            self.assertEqual(result["status"], "success")
            self.assertEqual(len(calls), 1)
            self.assertEqual(
                calls[0]["url"],
                "http://capability.local/api/internal/capability/migrations/apply",
            )
            headers = calls[0]["headers"]
            self.assertIsInstance(headers, dict)
            self.assertEqual(headers.get("Authorization"), "Bearer ephemeral-token")

            payload = json.loads(str(calls[0]["body"]))
            self.assertEqual(payload["migration_version"], "20260329_01_init")
            self.assertEqual(payload["sql"], "SELECT 1;\n")
        finally:
            shutil.rmtree(root, ignore_errors=True)


if __name__ == "__main__":
    unittest.main()

File: typescript/src/migration.test.ts
import assert from "node:assert/strict"
import { mkdtempSync, mkdirSync, rmSync, writeFileSync } from "node:fs"
import { tmpdir } from "node:os"
import path from "node:path"
import { describe, it } from "node:test"

import { migrate } from "./migration"

describe("migrate", () => {
  it("posts declared migration SQL to capability API", async () => {
    const root = mkdtempSync(path.join(tmpdir(), "octoarms-sdk-migrate-"))
    const migrationsDir = path.join(root, "migrations")
    mkdirSync(migrationsDir)
    writeFileSync(path.join(migrationsDir, "20260329_01_init.sql"), "SELECT 1;\n", "utf-8")

    const calls: Array<{ url: string; method: string; headers: Record<string, string>; body: string }> = []
    try {
      const result = await migrate(
        {
          taskName: "defillama_tvl_top10",
          taskVersion: "v0.1.0",
          migrationVersion: "20260329_01_init",
          capabilityEndpoint: "http://capability.local",
          ephemeralToken: "ephemeral-token",
          migrationIndex: { "20260329_01_init": "migrations/20260329_01_init.sql" },
          taskRootDir: root,
        },
        async (url, method, headers, body) => {
          calls.push({ url, method, headers, body })
          return { status: 200, text: '{"code":0,"msg":"success"}' }
        },
      )

      assert.equal(result.status, "success")
      assert.equal(calls.length, 1)
      assert.equal(calls[0]?.url, "http://capability.local/api/internal/capability/migrations/apply")
      assert.equal(calls[0]?.headers["Authorization"], "Bearer ephemeral-token")
    } finally {
      rmSync(root, { recursive: true, force: true })
    }
  })
})
